packages/core/src/anchor/auth/did-anchor-service-auth.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  AnchorServiceAuth,
+  CeramicApi,
+  DiagnosticsLogger,
+  FetchOpts,
+  HttpMethods,
+  fetchJson
+} from "@ceramicnetwork/common"
+import { DagJWS } from "dids"
+
+export class DIDAnchorServiceAuth implements AnchorServiceAuth {
+  private _ceramic: CeramicApi
+  private _nonce: number
+  private readonly _anchorServiceUrl: string
+  private readonly _logger: DiagnosticsLogger
+
+  constructor(
+    anchorServiceUrl: string,
+    logger: DiagnosticsLogger,
+  ) {
+    this._anchorServiceUrl = anchorServiceUrl
+    this._logger = logger
+  }
+
+  /**
+   * Set Ceramic API instance
+   *
+   * @param ceramic - Ceramic API used for various purposes
+   */
+  set ceramic(ceramic: CeramicApi) {
+    this._ceramic = ceramic
+  }
+
+  get nonce(): number {
+    return this._nonce
+  }
+
+  init = async (): Promise<void> => {
+    this._nonce = await this.lookupLastNonce() ?? 0
+  }
+
+  lookupLastNonce = async (): Promise<number | undefined> => {
+    if (!this._ceramic) {
+      throw new Error('Missing Ceramic instance required by this auth method')
+    }
+    let nonce
+    const { authorization } = await this.signRequest(this._getNonceEndpoint())
+    const data = await this._sendRequest(this._getNonceEndpoint(), {
+      method: HttpMethods.POST,
+      headers: { authorization },
+      body: {did: this._ceramic.did.id}
+    })
+    if (data) {
+      if (data.nonce) {
+        nonce = Number(data.nonce)
+        if (!isNaN(nonce)) {
+          return nonce
+        }
+      }
+    }
+    this._logger.debug('Could not get nonce from anchor service')
+    return
+  }
+
+  sendAuthenticatedRequest = async (url: URL | string, opts?: FetchOpts): Promise<any> => {
+    if (!this._ceramic) {
+      throw new Error('Missing Ceramic instance required by this auth method')
+    }
+    this._updateNonce()
+    const { authorization } = await this.signRequest(url, opts?.body, this._nonce)
+    let headers: any = { authorization }
+    opts?.headers && (headers = {...opts.headers, headers })
+    return await this._sendRequest(url, { ...opts, headers})
+  }
+
+  signRequest = async (url: URL | string, body?: any, nonce?: number): Promise<{jws: DagJWS, authorization: string}> => {
+    let payload: any = { url }
+    body && (payload = { ...payload, body})
+    nonce && (payload = { ...payload, nonce})
+    const jws = await this._ceramic.did.createJWS(payload)
+    const authorization = `Basic ${jws.signatures[0].protected}.${jws.payload}.${jws.signatures[0].signature}`
+    return { jws, authorization }
+  }
+
+  /**
+   * Increments nonce in memory.
+   */
+  private _updateNonce = async (): Promise<void> => {
+    this._nonce = this._nonce + 1
+  }
+
+  private _sendRequest = async (url: URL| string, opts?: FetchOpts): Promise<any> => {
```

Comment:
good point. I will add tests for this. the idea I have for this is, from ceramic every request must increment the nonce. skipping nonces (e.g. send 1 then send 3) is fine, repeating a nonce (e.g. send 1 then send 1 again) is never ok.

CAS is responsible for storing the nonce whether the request fails or not. if storing the nonce fails then CAS is broken.

Reply:
the nonce is used to prevent replay attack and identify unique requests. using an expiry timestamp would not allow replay attack to be prevented. unless each timestamp must be unique. a bit tough depending on precision but also might as well use a nonce in that case.

I think the risk of replay attack is small to be fair but taking similar precaution as with the admin api (except in this case I don't want to do a challenge response because that makes unnecessary extra requests)

Reply:
Another approach might be to use a random nonce? This would remove the need to make an additional request to retrieve a nonce and the work of keeping track of it on the js-ceramic side.

Also FYI, @ukstv's work on CAS w/o Ceramic node will add a timestamp to the request regardless.

Reply:
between challenge-response like the AdminAPI does, timestamp, or random nonce, I have no particular strong feelings.  I do think the incrementing nonce approach though does add extra fragility to the system, so I think one of the other 3 would be better if it can achieve the same goals

Reply:
How does random nonce work @oed ?

That sounds like the same as challenge response or is it different?
Or is it just that you can pick any number as long as you haven't used it before? That means we have to track every nonce ever used.

Reply:
I'll be updating this to the random nonce implementation and the nonces will be tracked by CAS